src/eliater/discover_latent_nodes.py

```diff
@@ -0,0 +1,145 @@
+"""This module contains methods to discover and mark nuisance nodes in a network.
+
+Given an acyclic directed mixed graph (ADMG), along with the treatment and the outcome
+of interest, certain observable variables can be regarded as nuisances. This
+classification arises because they do not have any impact on the outcome and should not
+be involved in the estimation of the treatment's effect on the outcome. These specific
+variables are descendants of the variables on all causal paths that are not ancestors of
+the outcome. A causal path, in this context, refers to a directed path that starts from the
+treatment and leads to the outcome such that all the arrows on the path have the same direction.
+This module is designed to identify these variables and produce a new ADMG in which they are
+designated as latent.
+
+This process enables us to concentrate on the fundamental variables needed to estimate the
+treatment's impact on the outcome. This focus results in more precise estimates with reduced
+variance and bias.
+
+Here is an example of an ADMG where X is the treatment and Y is the outcome is Y. This ADMG has
+only one causal path from X to Y which is X -> M1 -> M2 -> Y. The descendants of these variables
+that are ancestors of the outcome are R1, R2, and R3. The goal of this example is to identify these
+nuisance variables and mark them as latent.
+
+.. code-block:: python
+
+    import eliater
+    import y0
+    from y0.algorithm.identify import identify_outcomes
+    from y0.dsl import Variable, X, Y
+    from y0.graph import NxMixedGraph
+    from eliater.discover_latent_nodes import mark_latent
+
+    M1 = Variable("M1")
+    M2 = Variable("M2")
+    R1 = Variable("R1")
+    R2 = Variable("R2")
+    R3 = Variable("R3")
+
+    graph = NxMixedGraph.from_edges(
+        directed=[
+            (X, M1),
+            (M1, M2),
+            (M2, Y),
+            (M1, R1),
+            (R1, R2),
+            (R2, R3),
+            (Y, R3),
+        ],
+        undirected=[
+            (X, Y),
+        ],
+    )
+
+    new_graph = mark_latent(graph, treatments = Variable("X"), outcome: Variable("Y"))
+
+    # FIXME some unknown magic happens
```

Comment:
todo

Reply:
what happens after you know which nodes are latent?
without making this explicitly clear both in the preamble above and in this example, this entire module doesn't make sense

Reply:
i.e., what code do I use next after marking my variables as latent before I go to make my estimation

Reply:
I fixed the typo in the code block above:

`new_graph = mark_latent(graph, treatments=Variable("X"), outcomes=Variable("Y"))`

By unknown magic, did you mean this typo? 

One more question. Why did you generate the estimand? 

I just figured out that in y0 the estimand will be the same before and after the process of marking nodes as latent. It seems like y0 is able to not include the nuisance variables from the beginning without identifying them. In this case, the second huge benefit of this process is the application of simplification rules. For example, imagine we have a huge network (last case study in the paper), by identifying and marking all the nuisance variables as latent, and then applying the simplification rules to this new graph, we can significantly reduce the network size,  making it visually more appealing.

Reply:
Hence, I think we should wait until the simplification rules are complete and then add a line of code for simplifying the graph and generating a new simplified version that does not contain the nuisance variables.

Reply:
This is good - you have figured out by running the code it doesn't do what you expect (this is why I have over and over again in this issue written to try running the code)

Consider that this PR is not complete until this actually does something. You can figured out how to use the following three functions to finish the job:

1. `y0.graph.NxMixedGraph.to_latent_variable_dag()`
2. `y0.algorithm.simplify_latent.simplify_latent_dag()` 
3. `y0.graph.NxMixedGraph.from_latent_variable_dag()`

Reply:
This explanation helped a lot. However one of the Evan's simplification rules is missed in Y0. The rules are as follows:

a. Remove latent variables with no children from the graph.
b. Transform a latent variable with parents into an exogenous variable where all its parents are connected to its children.
c. Remove an exogenous latent variable that has at most one child.
d. Remove redundant latent nodes.

Rule c is not yet in y0. Is it OK if we create a branch in y0 for implementing it?  
Then we can have these three functions to finish the job.

Reply:
Yes. Split the effort though, you can finish the implementation in this PR using the y0 functions, assuming they will do what's expected.

Reply:
I finished the implementation. However, the code I put in the doc string is not doing what we expect it to do. It finds the nuisance variables correctly. Converts the ADMG into LV DAG and set nuisance variables as latent in the LV DAG correctly. But the `simplify_latent_dag ` function is not performing correctly. I expect all the nuisance variables be removed from the graph after applying Evan's simplification rules, but they all stay there. Here is the code:

 ```
   from y0.algorithm.identify import identify_outcomes
    from y0.dsl import Variable, X, Y
    from y0.graph import NxMixedGraph, set_latent
    from y0.algorithm.simplify_latent import simplify_latent_dag
    from eliater.discover_latent_nodes import find_nuisance_variables

    M1 = Variable("M1")
    M2 = Variable("M2")
    R1 = Variable("R1")
    R2 = Variable("R2")
    R3 = Variable("R3")

    graph = NxMixedGraph.from_edges(
        directed=[
            (X, M1),
            (M1, M2),
            (M2, Y),
            (M1, R1),
            (R1, R2),
            (R2, R3),
            (Y, R3),
        ],
        undirected=[
            (X, Y),
        ],
    )
    
    nuisance_variables = find_nuisance_variables(graph, treatments=X, outcomes=Y)

    lv_dag = NxMixedGraph.to_latent_variable_dag(graph)

    set_latent(lv_dag, nuisance_variables)  # set the nuisance variables as latent

    simplified_lv_dag = simplify_latent_dag(lv_dag)

    #convert the simplified latent variable dag to a NxMixedGraph object
    new_graph = NxMixedGraph.from_latent_variable_dag(simplified_lv_dag.graph)
```

Reply:
A good place to start would be writing this as a unit test

Reply:
You might want to just copy/paste all of the code from https://github.com/y0-causal-inference/y0/blob/main/src/y0/algorithm/simplify_latent.py into a new module in eliater then modify it directly in this PR